File: src/pty.rs
use std::os::fd::{ IntoRawFd, RawFd };
use std::ffi::CString;
use std::io::{ BufReader, Read };
use std::thread;
use std::os::fd::FromRawFd;

use nix::fcntl::OFlag;
use nix::ioctl_write_int;
use nix::pty::ForkptyResult;
use nix::{ fcntl::{ fcntl, F_SETFL }, pty::forkpty, unistd::execv, sys::termios::Termios };
use winit::event_loop::EventLoopProxy;
use winit::window::Window;

use crate::ansi_parser::{ AnsiParser, AnsiSimdParser };
use crate::CustomEvent;

pub struct Pty {
    fd: RawFd,
}

impl Pty {
    pub fn new(
        proxy: EventLoopProxy<CustomEvent>,
        mut parser: AnsiSimdParser,
        rows: usize,
        cols: usize
    ) -> Self {
        let winsize = nix::pty::Winsize {
            ws_row: rows as u16,
            ws_col: cols as u16,
            ws_xpixel: 0,
            ws_ypixel: 0,
        };

        let fork_result = (unsafe { forkpty(Some(&winsize), None) }).expect("Failed to fork pty");

        let master_fd = match fork_result {
            ForkptyResult::Parent { master, .. } => master.into_raw_fd(),
            ForkptyResult::Child { .. } => {
                // Child process: Execute the shell
                let shell = CString::new("/bin/zsh").expect("CString::new failed");
                let args = [shell.clone(), CString::new("-l").unwrap()];
                execv(&shell, &args).expect("execv failed");
                // If execv succeeds, the following lines won't be executed
                -1
            }
        };

        // Only the parent continues beyond this point
        let flags = OFlag::O_RDWR | OFlag::O_NONBLOCK;
        fcntl(master_fd, F_SETFL(flags)).expect("Failed to set flags");
        std::env::set_var("TERM", "xterm-256color");

        // Spawn a thread to read from the pty
        thread::spawn(move || {
            // Larger buffer size - 1MB
            let mut buf = [0u8; 1024 * 1024];

            loop {
                // Batch read multiple times before parsing
                let mut total_read = 0;
                for _ in 0..4 {
                    // Try up to 4 reads before processing
                    unsafe {
                        let n = libc::read(
                            master_fd,
                            buf.as_mut_ptr().add(total_read) as *mut _,
                            buf.len() - total_read
                        );

                        match n {
                            0 => {
                                return;
                            } // EOF
                            n if n > 0 => {
                                total_read += n as usize;
                                if total_read >= buf.len() / 2 {
                                    break; // Buffer getting full, process it
                                }
                            }
                            _ => {
                                // if *libc::__errno_location() == libc::EAGAIN {
                                //     break; // No more data available right now
                                // }
                                // return;
                            }
                        }
                    }
                }

                if total_read > 0 {
                    let data = &buf[..total_read];
                    parser.parse(data);
                    proxy.send_event(CustomEvent::RequestRedraw).expect("Failed to send event");
                }
            }
        });

        Pty {
            fd: master_fd,
            // buffer,
        }
    }

    pub fn resize(&self, rows: usize, cols: usize) {
        let winsize = nix::pty::Winsize {
            ws_row: rows as u16,
            ws_col: cols as u16,
            ws_xpixel: 0,
            ws_ypixel: 0,
        };

        unsafe {
            libc::ioctl(self.fd, libc::TIOCSWINSZ, &winsize as *const _);
        }
    }

    pub fn write(&self, data: &[u8]) {
        unsafe {
            libc::write(self.fd, data.as_ptr() as *const _, data.len());
        }
    }
}

File: mtCore/pty.cpp
//
//  Pty.cpp
//  mt
//
//  Created by Mano Rajesh on 12/9/25.
//

#include "pty.h"
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <signal.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>
#include <util.h>
#include <fcntl.h>
#include <errno.h>

Pty::Pty() {}

Pty::~Pty() {
    stop();
}

void Pty::stop() {
    // Stop source first so no more handler calls happen.
    if (ptySource) {
        dispatch_source_cancel(ptySource);
        // cancel handler will close FD
        ptySource = nullptr;
    } else if (master_fd >= 0) {
        close(master_fd);
        master_fd = -1;
    }
    
    // Reap child (and kill process group if still running).
    if (shell_pid > 0) {
        // Send SIGHUP or SIGTERM first; SIGTERM is fine.
        kill(-shell_pid, SIGTERM);
        usleep(100000);
        kill(-shell_pid, SIGKILL);
        
        // Best-effort reap (avoid zombies)
        int status = 0;
        (void)waitpid(shell_pid, &status, WNOHANG);
        shell_pid = -1;
    }
}

static void set_cloexec(int fd) {
    int flags = fcntl(fd, F_GETFD);
    if (flags >= 0) (void)fcntl(fd, F_SETFD, flags | FD_CLOEXEC);
}

static void set_nonblocking(int fd) {
    int flags = fcntl(fd, F_GETFL);
    if (flags >= 0) (void)fcntl(fd, F_SETFL, flags | O_NONBLOCK);
}

bool Pty::start(int rows, int cols) {
    int slave_fd = -1;
    termios t{};
    winsize w{};
    w.ws_row = (unsigned short)(rows > 0 ? rows : 24);
    w.ws_col = (unsigned short)(cols > 0 ? cols : 80);
    
    if (openpty(&master_fd, &slave_fd, nullptr, &t, &w) == -1) {
        perror("openpty");
        return false;
    }
    
    set_cloexec(master_fd);
    // With dispatch_source, nonblocking is recommended (handler can be called spuriously).
    set_nonblocking(master_fd);
    
    pid_t pid = fork();
    if (pid < 0) {
        perror("fork");
        close(master_fd); master_fd = -1;
        close(slave_fd);
        return false;
    }
    
    if (pid == 0) {
        // CHILD
        close(master_fd);
        
        // New session, controlling terminal
        setsid();
        ioctl(slave_fd, TIOCSCTTY, 0);
        
        const char* home = getenv("HOME");
        if (home) chdir(home);
        
        setenv("TERM", "xterm-256color", 1);
        setenv("COLORTERM", "truecolor", 1);
        setenv("TERM_PROGRAM", "mt", 1);
        setenv("TERM_PROGRAM_VERSION", "1.0", 1);
        
        // Make the child the leader of its process group (so parent can kill -pid)
        setpgid(0, 0);
        
        dup2(slave_fd, STDIN_FILENO);
        dup2(slave_fd, STDOUT_FILENO);
        dup2(slave_fd, STDERR_FILENO);
        if (slave_fd > STDERR_FILENO) close(slave_fd);
        
        // Env that many programs expect
        setenv("TERM", "xterm-256color", 1);
        setenv("COLORTERM", "truecolor", 1);
        
        // Prefer $SHELL if you want:
        const char* shell = getenv("SHELL");
        if (!shell) shell = "/bin/zsh";
        execlp(shell, shell, "-l", (char*)nullptr);
        
        perror("execl");
        _exit(127);
    }
    
    // PARENT
    shell_pid = pid;
    close(slave_fd);
    
    // Create a serial queue: THIS is now your PTY+parser thread.
    // Do parsing here to avoid thread-hops.
    dispatch_queue_t q =
    dispatch_queue_create("com.mt.ptyqueue", DISPATCH_QUEUE_SERIAL);
    
    ptySource = dispatch_source_create(DISPATCH_SOURCE_TYPE_READ, master_fd, 0, q);
    if (!ptySource) {
        fprintf(stderr, "dispatch_source_create failed\n");
        close(master_fd); master_fd = -1;
        return false;
    }
    
    // IMPORTANT: avoid capturing raw `this` if object might die before cancel finishes.
    // If your Pty truly lives for app lifetime, it's fine, but safer:
    Pty* self = this;
    
    dispatch_source_set_event_handler(ptySource, ^{
        // Drain until EAGAIN so we don't miss data.
        for (;;) {
            uint8_t buf[16384];
            ssize_t n = read(self->master_fd, buf, sizeof(buf));
            if (n > 0) {
                if (self->onOutput) {
                    // SPEC CHANGE: onOutput should DO THE PARSING here (same queue/thread).
                    self->onOutput((const char*)buf, (size_t)n);
                }
                continue;
            }
            if (n == 0) {
                printf("[mtCore] child exited (EOF)\n");
                dispatch_source_cancel(self->ptySource);
                break;
            }
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                break;
            }
            if (errno == EINTR) {
                continue;
            }
            perror("[mtCore] read");
            dispatch_source_cancel(self->ptySource);
            break;
        }
    });
    
    dispatch_source_set_cancel_handler(ptySource, ^{
        if (self->master_fd >= 0) {
            close(self->master_fd);
            self->master_fd = -1;
        }
        printf("[mtCore] PTY closed\n");
    });
    
    dispatch_resume(ptySource);
    return true;
}

bool Pty::setWinSize(int rows, int cols, int pxW, int pxH) {
    if (master_fd < 0) return false;
    winsize w{};
    w.ws_row = (unsigned short)rows;
    w.ws_col = (unsigned short)cols;
    w.ws_xpixel = (unsigned short)pxW;
    w.ws_ypixel = (unsigned short)pxH;
    if (ioctl(master_fd, TIOCSWINSZ, &w) != 0) return false;
    if (shell_pid > 0) (void)kill(shell_pid, SIGWINCH);
    return true;
}

void Pty::send(const char *data, size_t len) {
    if (master_fd < 0 || !data || len == 0) return;
    
    size_t off = 0;
    while (off < len) {
        ssize_t n = write(master_fd, data + off, len - off);
        if (n > 0) { off += (size_t)n; continue; }
        if (n < 0 && errno == EINTR) continue;
        if (n < 0 && (errno == EAGAIN || errno == EWOULDBLOCK)) continue; // or yield/sleep
        break;
    }
}
